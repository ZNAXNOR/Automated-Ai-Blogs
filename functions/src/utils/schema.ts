import { z } from 'zod';

// Round 0: Trends
export const TrendItemSchema = z.object({
  query: z.string(),
  type: z.enum(['autocomplete', 'related', 'trending', 'rss']),
  score: z.number(),
  source: z.array(z.string()),
  reason: z.string().optional(),
});

export const Round0InputSchema = z.object({
  runId: z.string(),
  seeds: z.array(z.string()),
  region: z.string().optional(),
  useLLM: z.boolean().optional(),
  force: z.boolean().optional(),
});

export const Round0OutputSchema = z.object({
  items: z.array(TrendItemSchema),
  cached: z.boolean(),
  sourceCounts: z.record(z.number()),
});

// Round 1: Ideate
export const IdeationItemSchema = z.object({
  trend: z.string(),
  idea: z.string(),
  variant: z.number(),
  source: z.literal('llm'),
});

export const Round1InputSchema = Round0OutputSchema;
export const Round1OutputSchema = z.object({
    items: z.array(IdeationItemSchema)
});

// Round 2: Outline
export const OutlineItemSchema = z.object({
    idea: z.string(),
    outline: z.array(z.string()),
});
export const Round2InputSchema = Round1OutputSchema;
export const Round2OutputSchema = z.object({
    items: z.array(OutlineItemSchema)
});

// Round 3: Draft
export const DraftItemSchema = z.object({
    idea: z.string(),
    draft: z.string(),
});
export const Round3InputSchema = Round2OutputSchema;
export const Round3OutputSchema = z.object({
    items: z.array(DraftItemSchema)
});

// Round 4: Polish
export const PolishedDraftItemSchema = z.object({
    idea: z.string(),
    polishedDraft: z.string(),
});
export const Round4InputSchema = Round3OutputSchema;
export const Round4OutputSchema = z.object({
    items: z.array(PolishedDraftItemSchema)
});

// Round 5: Meta
export const MetaItemSchema = z.object({
    idea: z.string(),
    meta: z.object({
        title: z.string(),
        description: z.string(),
        keywords: z.array(z.string()),
    }),
});
export const Round5InputSchema = Round4OutputSchema;
export const Round5OutputSchema = z.object({
    items: z.array(MetaItemSchema)
});


// Round 6: Coherence
export const CoherenceItemSchema = z.object({
    idea: z.string(),
    coherenceScore: z.number(),
});
export const Round6InputSchema = Round5OutputSchema;
export const Round6OutputSchema = z.object({
    items: z.array(CoherenceItemSchema)
});

// Round 7: Publish
export const PublishItemSchema = z.object({
    idea: z.string(),
    url: z.string(),
});
export const Round7InputSchema = Round6OutputSchema;
export const Round7OutputSchema = z.object({
    items: z.array(PublishItemSchema)
});

export type TrendItem = z.infer<typeof TrendItemSchema>;
export type Round0Input = z.infer<typeof Round0InputSchema>;
export type Round0Output = z.infer<typeof Round0OutputSchema>;
export type IdeationItem = z.infer<typeof IdeationItemSchema>;
export type Round1Input = z.infer<typeof Round1InputSchema>;
export type Round1Output = z.infer<typeof Round1OutputSchema>;
export type OutlineItem = z.infer<typeof OutlineItemSchema>;
export type Round2Input = z.infer<typeof Round2InputSchema>;
export type Round2Output = z.infer<typeof Round2OutputSchema>;
export type DraftItem = z.infer<typeof DraftItemSchema>;
export type Round3Input = z.infer<typeof Round3InputSchema>;
export type Round3Output = z.infer<typeof Round3OutputSchema>;
export type PolishedDraftItem = z.infer<typeof PolishedDraftItemSchema>;
export type Round4Input = z.infer<typeof Round4InputSchema>;
export type Round4Output = z.infer<typeof Round4OutputSchema>;
export type MetaItem = z.infer<typeof MetaItemSchema>;
export type Round5Input = z.infer<typeof Round5InputSchema>;
export type Round5Output = z.infer<typeof Round5OutputSchema>;
export type CoherenceItem = z.infer<typeof CoherenceItemSchema>;
export type Round6Input = z.infer<typeof Round6InputSchema>;
export type Round6Output = z.infer<typeof Round6OutputSchema>;
export type PublishItem = z.infer<typeof PublishItemSchema>;
export type Round7Input = z.infer<typeof Round7InputSchema>;
export type Round7Output = z.infer<typeof Round7OutputSchema>;