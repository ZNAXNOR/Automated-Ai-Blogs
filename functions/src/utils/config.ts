import * as dotenv from 'dotenv';
import * as path from 'path';

dotenv.config({ path: path.resolve(process.cwd(), '.env.local'), override: true });

// --- Environment Validation ---
const requiredEnvVars = [
    'SERPAPI_KEY',
    'HF_TOKEN',
    'WP_API_URL',
    'WP_USERNAME',
    'WP_PASSWORD'
];
const missingVars = requiredEnvVars.filter(varName => !process.env[varName]);
if (missingVars.length > 0) {
    throw new Error(`Missing required environment variables: ${missingVars.join(', ')}`);
}

// --- Format Validation ---
try {
    new URL(process.env.WP_API_URL!);
} catch (e) {
    throw new Error('WP_API_URL is not a valid URL.');
}

if (process.env.CACHE_TTL_HOURS) {
    const ttl = Number(process.env.CACHE_TTL_HOURS);
    if (isNaN(ttl) || ttl <= 0) {
        throw new Error('CACHE_TTL_HOURS must be a positive number.');
    }
}

// --- Exported Configuration ---
export const env = {
    // API Keys
    serpApiKey: process.env.SERPAPI_KEY!,
    hfToken: process.env.HF_TOKEN!,

    // WordPress
    wpApiUrl: process.env.WP_API_URL!,
    wpUsername: process.env.WP_USERNAME!,
    wpPassword: process.env.WP_PASSWORD!,

    // Feature Flags & Settings
    useR0Llm: process.env.USE_R0_LLM === 'true',
    CACHE_TTL_HOURS: Number(process.env.CACHE_TTL_HOURS) || 24,

    // Model Slugs 
    hfModelR1: 'TinyLlama/TinyLlama-1.1B-Chat-v1.0',
    hfModelR2: 'microsoft/Phi-3-mini-4k-instruct',
    hfModelR3: 'mistralai/Mistral-7B-Instruct-v0.2',
    hfModelR4: 'mistralai/Mistral-7B-Instruct-v0.2',
    hfModelR5: 'google/gemma-2-7b-it',
    hfModelR6: 'sentence-transformers/all-MiniLM-L6-v2',
};
