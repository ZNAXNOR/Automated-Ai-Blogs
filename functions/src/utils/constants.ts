import { z } from "zod";

const minute = 60;
const hour = 60 * minute;

export const constants = {
    // --- Round 0 --- 
    TRENDS_PER_RUN: 25,
    TINY_HF_MODEL: "gpt2",

    // --- Round 3 ---
    MIN_DRAFT_WORD_COUNT: 250,
    MAX_DRAFT_WORD_COUNT: 400,

    // --- Round 4 ---
    MIN_POLISH_WORD_COUNT: 250,
    MAX_POLISH_WORD_COUNT: 500,
    MAX_REFINEMENT_ATTEMPTS: 2,

    // --- Round 5 ---
    MIN_META_WORD_COUNT: 100,

    // --- Round 7 ---
    DEFAULT_WP_TAGS: ["AI Content"],

    // --- Global ---
    MAX_RUN_DURATION_SECONDS: 20 * minute,
    MAX_RETRY_ATTEMPTS: 2,
    RUNS_COLLECTION: "runs",

    // --- Firestore Paths ---
    ARTIFACT_PATHS: {
        R0_TRENDS: 'runs/{runId}/artifacts/round0/data',
        R1_IDEAS: 'runs/{runId}/artifacts/round1/data',
        R2_OUTLINES: 'runs/{runId}/artifacts/round2/data',
        R3_DRAFTS: 'runs/{runId}/artifacts/round3/data',
        R4_POLISHED: 'runs/{runId}/artifacts/round4/data',
        R4_FAILURES: 'runs/{runId}/artifacts/round4-failures/data',
        R5_METADATA: 'runs/{runId}/artifacts/round5/data',
        R6_COHERENCE: 'runs/{runId}/artifacts/round6/data',
        R7_PUBLISHED: 'runs/{runId}/artifacts/round7/data',
    },

    // --- Caching ---
    CACHE_TTL: {
        DEFAULT: 1 * hour,
        SERP: 24 * hour,
        TRENDS: 24 * hour,
        IDEAS: 1 * hour,
        DRAFTS: 1 * hour,
        METADATA: 1 * hour,
        COHERENCE: 1 * hour,
    }

};

export const TrendItem = z.object({
    query: z.string(),
    type: z.enum(["trending", "related", "autocomplete", "rss"]),
    score: z.number(),
    source: z.array(z.string()),
});
