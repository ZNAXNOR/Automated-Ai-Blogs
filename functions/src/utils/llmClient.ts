
export class LLMClient {
  async generate(options: {
    model: string;
    prompt: string;
    max_tokens: number;
    temperature: number;
  }): Promise<{ text: string }> {
    // This is a placeholder and will be mocked in tests.
    return { text: "" };
  }
}
