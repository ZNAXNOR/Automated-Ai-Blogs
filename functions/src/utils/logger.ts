/**
 * Logger utility for structured logging.
 */
const log = (level: 'INFO' | 'WARN' | 'ERROR', message: string, data?: any) => {
    const logObject = {
        level,
        message,
        timestamp: new Date().toISOString(),
        ...(data && { data }),
    };
    console.log(JSON.stringify(logObject));
};

export const logger = {
    info: (message: string, data?: any) => log('INFO', message, data),
    warn: (message: string, data?: any) => log('WARN', message, data),
    error: (message: string, data?: any) => log('ERROR', message, data),
};
