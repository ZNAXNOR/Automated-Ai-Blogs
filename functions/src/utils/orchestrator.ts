import * as R0 from '../rounds/r0_trends';
import * as R1 from '../rounds/r1_ideate';
import * as R2 from '../rounds/r2_outline';
import * as R3 from '../rounds/r3_draft';
import * as R4 from '../rounds/r4_polish';
import * as R5 from '../rounds/r5_meta';
import * as R6 from '../rounds/r6_coherence';
import * as R7 from '../rounds/r7_publish';

import { JobPayload } from './types';

interface PipelineResult {
  payload: JobPayload;
  status: 'completed' | 'failed';
  error?: string;
  wpPostUrl?: string;
  wpPostId?: number;
}

export async function runPipeline(initialTrendInput: any): Promise<PipelineResult> {
  const payload: JobPayload = {
    trendInput: initialTrendInput
  };

  try {
    const out0 = await R0.run(payload);
    Object.assign(payload, out0);

    const out1 = await R1.run(payload);
    Object.assign(payload, out1);

    const out2 = await R2.run(payload);
    Object.assign(payload, out2);

    const out3 = await R3.run(payload);
    Object.assign(payload, out3);

    const out4 = await R4.run(payload);
    Object.assign(payload, out4);

    const out5 = await R5.run(payload);
    Object.assign(payload, out5);

    const out6 = await R6.run(payload);
    Object.assign(payload, out6);

    const out7 = await R7.run(payload);
    Object.assign(payload, { wpPostUrl: out7.wpPostUrl, wpPostId: out7.wpPostId });

    return {
      payload,
      status: 'completed',
      wpPostUrl: out7.wpPostUrl,
      wpPostId: out7.wpPostId
    };
  } catch (err: any) {
    return {
      payload,
      status: 'failed',
      error: err.message
    };
  }
}
