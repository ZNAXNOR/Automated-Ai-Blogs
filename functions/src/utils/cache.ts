/**
 * Simple Firestore cache with TTL.
 * Documents under collection `cache` with nested keys like `serpapi/{hash}`.
 *
 * Shape:
 *  {
 *    payload: any,
 *    createdAt: Timestamp,
 *    expiresAt: Timestamp
 *  }
 */

import * as admin from "firebase-admin";
import { env } from "./config";

if (!admin.apps.length) {
  admin.initializeApp();
}
const db = admin.firestore();

export async function getCache(pathKey: string): Promise<{ payload: any } | null> {
  // pathKey example: "serpapi:abcd" -> doc path "cache/serpapi:abcd"
  const ref = db.doc(`cache/${pathKey}`);
  const snap = await ref.get();
  if (!snap.exists) return null;
  const data = snap.data()!;
  const expiresAt = data.expiresAt?.toDate?.() ?? new Date(0);
  if (expiresAt.getTime() < Date.now()) return null;
  return { payload: data.payload };
}

export async function setCache(pathKey: string, payload: any, ttlHours?: number): Promise<void> {
  const hours = (ttlHours ?? env.CACHE_TTL_HOURS ?? 24);
  const now = new Date();
  const expires = new Date(now.getTime() + hours * 3600 * 1000);
  const ref = db.doc(`cache/${pathKey}`);
  await ref.set({
    payload,
    createdAt: admin.firestore.FieldValue.serverTimestamp(),
    expiresAt: admin.firestore.Timestamp.fromDate(expires),
  });
}
