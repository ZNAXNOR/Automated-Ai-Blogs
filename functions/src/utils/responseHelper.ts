import { z } from 'zod';
import { logger } from './logger';
import type { Response } from 'node-fetch';

/**
 * A wrapper for fetch responses that provides normalized .json() and .text() methods.
 */
export class ResponseWrapper {
    private constructor(private response: Response) {}

    static create(response: Response): ResponseWrapper {
        return new ResponseWrapper(response);
    }

    async json<T>(schema: z.ZodType<T>): Promise<T> {
        const text = await this.response.text();
        try {
            // Extract JSON from markdown code blocks
            let jsonText = text.trim();
            if (jsonText.startsWith('```json')) {
                jsonText = jsonText.substring(7);
            }
            if (jsonText.endsWith('```')) {
                jsonText = jsonText.slice(0, -3);
            }

            let data = JSON.parse(jsonText);
            // Handle double-stringified JSON
            if (typeof data === 'string') {
                data = JSON.parse(data);
            }
            const validatedData = schema.parse(data);
            return validatedData;
        } catch (error) {
            logger.error('ResponseWrapper.json: Validation failed', { error, text });
            throw error;
        }
    }

    async text(): Promise<string> {
        return this.response.text();
    }
}
