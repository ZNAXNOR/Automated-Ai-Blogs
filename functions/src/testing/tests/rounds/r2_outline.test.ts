/**
 * Unit & Integration Tests for Round 2 (Outline Generation)
 */

import {
  Round2_Outline,
  fetchR1Data,
  saveR2Outlines,
  _test,
} from "../../../rounds/r2_outline";
import { IdeationItem } from "../../../utils/schema";
import fetch, { Response } from "node-fetch";

// --- MOCKS ---
const setMock = jest.fn();
const getMock = jest.fn();
const docMock = jest.fn(() => ({
  get: getMock,
  set: setMock,
}));

jest.mock("firebase-admin/firestore", () => ({
  getFirestore: jest.fn(() => ({ doc: docMock })),
}));

jest.mock("node-fetch", () => jest.fn());
const mockedFetch = fetch as unknown as jest.MockedFunction<typeof fetch>;

const { buildPrompt, extractJsonFromText, callHuggingFace, validateIdeationItems, validateOutlineSchema } = _test;

// --- TEST DATA ---
const RUN_ID = "test-run-r2";
const R1_ITEMS: IdeationItem[] = [
  { trend: "AI in healthcare", idea: "AI in hospitals", variant: 1, source: "llm" },
  { trend: "Budget smartphones", idea: "Top cheap phones", variant: 1, source: "llm" },
];

const LLM_RAW_OUTPUT = `
  Here is the JSON you requested:
  [
    {
      "trend": "AI in healthcare",
      "idea": "AI in hospitals",
      "sections": [
        { "heading": "Intro", "bullets": ["Point A"], "estWordCount": 50 },
        { "heading": "Conclusion", "bullets": ["Point B"], "estWordCount": 50 }
      ]
    },
    {
      "trend": "Budget smartphones",
      "idea": "Top cheap phones",
      "sections": [
        { "heading": "Intro", "bullets": ["C"], "estWordCount": 50 },
        { "heading": "Body", "bullets": ["D"], "estWordCount": 150 }
      ]
    }
  ]
  Hope you find it useful!
`;

const PARSED_OUTLINES = [
  {
    trend: "AI in healthcare",
    idea: "AI in hospitals",
    sections: [
      { heading: "Intro", bullets: ["Point A"], estWordCount: 50 },
      { heading: "Conclusion", bullets: ["Point B"], estWordCount: 50 },
    ],
  },
  {
    trend: "Budget smartphones",
    idea: "Top cheap phones",
    sections: [
      { heading: "Intro", bullets: ["C"], estWordCount: 50 },
      { heading: "Body", bullets: ["D"], estWordCount: 150 },
    ],
  },
];

// --- TESTS ---
describe("Round 2: Outline Generation", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    process.env.HUGGINGFACE_API_KEY = "test-key";
    process.env.HUGGINGFACE_MODEL_R2 = "test-model-r2";
  });

  // 1. Test I/O Functions
  describe("I/O Operations", () => {
    it("fetchR1Data should retrieve and validate data", async () => {
      getMock.mockResolvedValue({ exists: true, data: () => ({ items: R1_ITEMS }) });
      const items = await fetchR1Data(RUN_ID);
      expect(docMock).toHaveBeenCalledWith(`runs/${RUN_ID}/artifacts/round1`);
      expect(getMock).toHaveBeenCalledTimes(1);
      expect(items).toEqual(R1_ITEMS);
    });

    it("fetchR1Data should throw if artifact not found", async () => {
      getMock.mockResolvedValue({ exists: false });
      await expect(fetchR1Data(RUN_ID)).rejects.toThrow(`R1 artifact not found for runId=${RUN_ID}`);
    });

    it("saveR2Outlines should write data to Firestore", async () => {
      await saveR2Outlines(RUN_ID, PARSED_OUTLINES);
      expect(docMock).toHaveBeenCalledWith(`runs/${RUN_ID}/artifacts/round2`);
      expect(setMock).toHaveBeenCalledWith({ items: PARSED_OUTLINES }, { merge: true });
    });
  });

  // 2. Test Utility Functions
  describe("Utility Functions", () => {
    it("buildPrompt should create a valid prompt", () => {
      const prompt = buildPrompt(R1_ITEMS);
      expect(prompt).toContain("AI in healthcare");
      expect(prompt).toContain("Top cheap phones");
      expect(prompt).toMatchSnapshot(); // Good for catching unintentional prompt changes
    });

    it("extractJsonFromText should get JSON from raw text", () => {
      const json = extractJsonFromText(LLM_RAW_OUTPUT);
      const expectedJson = JSON.stringify(PARSED_OUTLINES, null, 2).replace(/\n/g, '\n  ');
      //This is a bit of a hack to get the formatting to match the snapshot.

      expect(JSON.parse(json!)).toEqual(PARSED_OUTLINES);
    });

    it("extractJsonFromText should return null if no JSON is found", () => {
      expect(extractJsonFromText("No JSON here")).toBeNull();
    });
  });

  // 3. Test Validation Logic
  describe("Validation", () => {
    it("validateIdeationItems should pass valid items", () => {
      expect(() => validateIdeationItems(R1_ITEMS)).not.toThrow();
    });

    it("validateIdeationItems should throw on invalid items", () => {
      const invalidItems = [...R1_ITEMS, { trend: "Missing fields" }];
      expect(() => validateIdeationItems(invalidItems)).toThrow(/Invalid IdeationItem/);
    });

    it("validateOutlineSchema should pass valid outlines", () => {
      expect(() => validateOutlineSchema(PARSED_OUTLINES)).not.toThrow();
    });

    it("validateOutlineSchema should throw on invalid outlines", () => {
      const invalidOutlines = [...PARSED_OUTLINES, { trend: "Missing sections" }];
      expect(() => validateOutlineSchema(invalidOutlines)).toThrow(/Invalid outline format/);
    });
     it("validateOutlineSchema should throw on invalid sections", () => {
      const invalidSection = JSON.parse(JSON.stringify(PARSED_OUTLINES));
      invalidSection[0].sections.push({ heading: "No bullets" });
      expect(() => validateOutlineSchema(invalidSection)).toThrow(/Invalid section format/);
    });
  });


  // 4. Test API Call Logic
  describe("Hugging Face API Call", () => {
    it("callHuggingFace should return text on success", async () => {
      const mockResponse = {
        ok: true,
        json: async () => [{ generated_text: "Success!" }],
      } as Response;
      mockedFetch.mockResolvedValue(mockResponse);

      const result = await callHuggingFace("test prompt");
      expect(result).toBe("Success!");
    });

    it("callHuggingFace should throw on API error", async () => {
      const mockResponse = { ok: false, status: 500, statusText: "Server Error", text: async()=>"Error" } as Response;
      mockedFetch.mockResolvedValue(mockResponse);
      await expect(callHuggingFace("test prompt")).rejects.toThrow(/Hugging Face API Error/);
    });

     it("callHuggingFace should throw on invalid response structure", async () => {
      const mockResponse = {
        ok: true,
        json: async () => ({ unexpected: "structure" }), // Malformed
      } as Response;
      mockedFetch.mockResolvedValue(mockResponse);
      await expect(callHuggingFace("test prompt")).rejects.toThrow(/Invalid or missing 'generated_text'/);
    });
  });


  // 5. Test Full Orchestration
  describe("Full Orchestration: Round2_Outline", () => {
    it("should execute the full round successfully", async () => {
      // Arrange
      getMock.mockResolvedValue({ exists: true, data: () => ({ items: R1_ITEMS }) });
      const mockResponse = {
        ok: true,
        json: async () => [{ generated_text: LLM_RAW_OUTPUT }],
      } as Response;
      mockedFetch.mockResolvedValue(mockResponse);

      // Act
      await Round2_Outline(RUN_ID);

      // Assert
      // 1. Fetched R1 data
      expect(docMock).toHaveBeenCalledWith(`runs/${RUN_ID}/artifacts/round1`);
      // 2. Called Hugging Face
      expect(mockedFetch).toHaveBeenCalledTimes(1);
      // 3. Saved R2 data
      expect(docMock).toHaveBeenCalledWith(`runs/${RUN_ID}/artifacts/round2`);
      expect(setMock).toHaveBeenCalledWith({ items: PARSED_OUTLINES }, { merge: true });
    });

     it("should retry on failed Hugging Face call", async () => {
      getMock.mockResolvedValue({ exists: true, data: () => ({ items: R1_ITEMS }) });

      const failureResponse = { ok: false, status: 504, statusText: "Gateway Timeout", text: async()=>"Timeout" } as Response;
      const successResponse = {
        ok: true,
        json: async () => [{ generated_text: LLM_RAW_OUTPUT }],
      } as Response;

      // Fail twice, then succeed
      mockedFetch.mockResolvedValueOnce(failureResponse).mockResolvedValueOnce(failureResponse).mockResolvedValueOnce(successResponse);

      await Round2_Outline(RUN_ID);

      expect(mockedFetch).toHaveBeenCalledTimes(3);
      expect(setMock).toHaveBeenCalledTimes(1); // Should still succeed
    });

  });
});
