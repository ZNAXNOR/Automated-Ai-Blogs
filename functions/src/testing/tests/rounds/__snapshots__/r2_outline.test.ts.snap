// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Round 2: Outline Generation Utility Functions buildPrompt should create a valid prompt 1`] = `
"
    Given a list of trends and corresponding content ideas, generate a detailed article outline for each.

    INPUT:
    [
  {
    "trend": "AI in healthcare",
    "idea": "AI in hospitals"
  },
  {
    "trend": "Budget smartphones",
    "idea": "Top cheap phones"
  }
]

    INSTRUCTIONS:
    For each idea, create an outline with these properties:
    - "trend": The original trend.
    - "idea": The original idea.
    - "sections": An array of objects, where each object has:
      - "heading": A descriptive heading for the section (e.g., "Introduction", "The Rise of AI in Healthcare").
      - "bullets": An array of strings, with each string being a key point or question to cover in that section.
      - "estWordCount": A number representing the estimated word count for the section.

    The final output MUST be a single, clean JSON array of outline objects. Do not include any text or formatting before or after the JSON.

    Example Output:
    [
      {
        "trend": "AI in healthcare",
        "idea": "How AI is revolutionizing patient diagnostics",
        "sections": [
          {
            "heading": "Introduction",
            "bullets": [
              "Brief overview of AI's growing role in medicine.",
              "Thesis: AI-powered diagnostics are improving accuracy and speed, leading to better patient outcomes."
            ],
            "estWordCount": 150
          },
          {
            "heading": "Conclusion",
            "bullets": [
              "Summary of key benefits.",
              "Future outlook and potential challenges."
            ],
            "estWordCount": 200
          }
        ]
      }
    ]
    "
`;
